### PR TITLE
Wip/hughsie/hid

### PR DESCRIPTION
**Pull request by hughsie**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [X] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by hughsie** [comment]
@GamePad64 -- it seems quite a lot of hardware doesn't want to return the HID descriptor on the alternate interface too -- e.g. this is `sudo lsusb -vv` with a Bolt dongle plugged in:

	Bus 005 Device 006: ID 046d:c548 Logitech, Inc. Logi Bolt Receiver
	Device Descriptor:
	  bLength                18
	  bDescriptorType         1
	  bcdUSB               2.00
	  bDeviceClass            0 [unknown]
	  bDeviceSubClass         0 [unknown]
	  bDeviceProtocol         0 
	  bMaxPacketSize0        64
	  idVendor           0x046d Logitech, Inc.
	  idProduct          0xc548 Logi Bolt Receiver
	  bcdDevice            5.00
	  iManufacturer           1 Logitech
	  iProduct                2 USB Receiver
	  iSerial                 0 
	  bNumConfigurations      1
	  Configuration Descriptor:
	    bLength                 9
	    bDescriptorType         2
	    wTotalLength       0x0054
	    bNumInterfaces          3
	    bConfigurationValue     1
	    iConfiguration          4 MPR05.00_B0008
	    bmAttributes         0xa0
	      (Bus Powered)
	      Remote Wakeup
	    MaxPower               98mA
	    Interface Descriptor:
	      bLength                 9
	      bDescriptorType         4
	      bInterfaceNumber        0
	      bAlternateSetting       0
	      bNumEndpoints           1
	      bInterfaceClass         3 Human Interface Device
	      bInterfaceSubClass      1 Boot Interface Subclass
	      bInterfaceProtocol      1 Keyboard
	      iInterface              0 
		HID Device Descriptor:
		  bLength                 9
		  bDescriptorType        33
		  bcdHID               1.11
		  bCountryCode            0 Not supported
		  bNumDescriptors         1
		  bDescriptorType        34 Report
		  wDescriptorLength      67
		  Report Descriptor: (length is 67)
		    Item(Global): Usage Page, data= [ 0x01 ] 1
				    Generic Desktop Controls
		    Item(Local ): Usage, data= [ 0x06 ] 6
				    Keyboard
		    Item(Main  ): Collection, data= [ 0x01 ] 1
				    Application
		    Item(Global): Usage Page, data= [ 0x07 ] 7
				    Keyboard
		    Item(Local ): Usage Minimum, data= [ 0xe0 ] 224
				    Control Left
		    Item(Local ): Usage Maximum, data= [ 0xe7 ] 231
				    GUI Right
		    Item(Global): Logical Minimum, data= [ 0x00 ] 0
		    Item(Global): Logical Maximum, data= [ 0x01 ] 1
		    Item(Global): Report Size, data= [ 0x01 ] 1
		    Item(Global): Report Count, data= [ 0x08 ] 8
		    Item(Main  ): Input, data= [ 0x02 ] 2
				    Data Variable Absolute No_Wrap Linear
				    Preferred_State No_Null_Position Non_Volatile Bitfield
		    Item(Global): Report Count, data= [ 0x05 ] 5
		    Item(Global): Usage Page, data= [ 0x08 ] 8
				    LEDs
		    Item(Local ): Usage Minimum, data= [ 0x01 ] 1
				    NumLock
		    Item(Local ): Usage Maximum, data= [ 0x05 ] 5
				    Kana
		    Item(Main  ): Output, data= [ 0x02 ] 2
				    Data Variable Absolute No_Wrap Linear
				    Preferred_State No_Null_Position Non_Volatile Bitfield
		    Item(Global): Report Count, data= [ 0x01 ] 1
		    Item(Global): Report Size, data= [ 0x03 ] 3
		    Item(Main  ): Output, data= [ 0x03 ] 3
				    Constant Variable Absolute No_Wrap Linear
				    Preferred_State No_Null_Position Non_Volatile Bitfield
		    Item(Global): Report Count, data= [ 0x70 ] 112
		    Item(Global): Report Size, data= [ 0x01 ] 1
		    Item(Global): Usage Page, data= [ 0x07 ] 7
				    Keyboard
		    Item(Local ): Usage Minimum, data= [ 0x04 ] 4
				    A
		    Item(Local ): Usage Maximum, data= [ 0x73 ] 115
				    F24
		    Item(Main  ): Input, data= [ 0x02 ] 2
				    Data Variable Absolute No_Wrap Linear
				    Preferred_State No_Null_Position Non_Volatile Bitfield
		    Item(Global): Report Count, data= [ 0x05 ] 5
		    Item(Local ): Usage Minimum, data= [ 0x87 ] 135
				    International 1 (PC98)
		    Item(Local ): Usage Maximum, data= [ 0x8b ] 139
				    International 5 (PC98)
		    Item(Main  ): Input, data= [ 0x02 ] 2
				    Data Variable Absolute No_Wrap Linear
				    Preferred_State No_Null_Position Non_Volatile Bitfield
		    Item(Global): Report Count, data= [ 0x03 ] 3
		    Item(Local ): Usage Minimum, data= [ 0x90 ] 144
				    LANG 1 (Hangul/English Toggle, Korea)
		    Item(Local ): Usage Maximum, data= [ 0x92 ] 146
				    LANG 3 (Katakana, Japan)
		    Item(Main  ): Input, data= [ 0x02 ] 2
				    Data Variable Absolute No_Wrap Linear
				    Preferred_State No_Null_Position Non_Volatile Bitfield
		    Item(Main  ): End Collection, data=none
	      Endpoint Descriptor:
		bLength                 7
		bDescriptorType         5
		bEndpointAddress     0x81  EP 1 IN
		bmAttributes            3
		  Transfer Type            Interrupt
		  Synch Type               None
		  Usage Type               Data
		wMaxPacketSize     0x0040  1x 64 bytes
		bInterval               1
	    Interface Descriptor:
	      bLength                 9
	      bDescriptorType         4
	      bInterfaceNumber        1
	      bAlternateSetting       0
	      bNumEndpoints           1
	      bInterfaceClass         3 Human Interface Device
	      bInterfaceSubClass      1 Boot Interface Subclass
	      bInterfaceProtocol      2 Mouse
	      iInterface              0 
		HID Device Descriptor:
		  bLength                 9
		  bDescriptorType        33
		  bcdHID               1.11
		  bCountryCode            0 Not supported
		  bNumDescriptors         1
		  bDescriptorType        34 Report
		  wDescriptorLength     133
		  Report Descriptors: 
		    ** UNAVAILABLE **
	      Endpoint Descriptor:
		bLength                 7
		bDescriptorType         5
		bEndpointAddress     0x82  EP 2 IN
		bmAttributes            3
		  Transfer Type            Interrupt
		  Synch Type               None
		  Usage Type               Data
		wMaxPacketSize     0x0040  1x 64 bytes
		bInterval               1
	    Interface Descriptor:
	      bLength                 9
	      bDescriptorType         4
	      bInterfaceNumber        2
	      bAlternateSetting       0
	      bNumEndpoints           1
	      bInterfaceClass         3 Human Interface Device
	      bInterfaceSubClass      0 [unknown]
	      bInterfaceProtocol      0 
	      iInterface              0 
		HID Device Descriptor:
		  bLength                 9
		  bDescriptorType        33
		  bcdHID               1.11
		  bCountryCode            0 Not supported
		  bNumDescriptors         1
		  bDescriptorType        34 Report
		  wDescriptorLength      54
		  Report Descriptors: 
		    ** UNAVAILABLE **

File: libfwupdplugin/fu-hid-device.c [hunk 1]
```diff
@@ -86,34 +86,40 @@ fu_hid_device_set_property(GObject *object, guint prop_id, const GValue *value,
...
  * @error: (nullable): optional return location for an error
  *
- * Parses the HID descriptor.
+ * Parses the HID descriptors.
  *
- * Returns: (transfer full): a #FuHidDescriptor, or %NULL for error
+ * Returns: (transfer container) (element-type FuHidDescriptor): descriptors, or %NULL for error
  *
- * Since: 1.9.4
+ * Since: 1.9.10
```
Comment (superm1) [inline libfwupdplugin/fu-hid-device.c:97]: Should be 1.9.11 or 2.0.0 now

File: plugins/cfu/fu-cfu-device.c [hunk 2]
```diff
@@ -404,19 +431,28 @@ fu_cfu_device_setup(FuDevice *device, GError **error)
...
+	descriptors = fu_hid_device_parse_descriptors(FU_HID_DEVICE(device), error);
+	if (descriptors == NULL)
 		return FALSE;
-	if (!fu_cfu_device_ensure_map_item(descriptor, &self->content_get_report, error))
+	for (guint i = 0; i < descriptors->len; i++) {
+		FuHidDescriptor *descriptor = g_ptr_array_index(descriptors, i);
+		g_autoptr(GError) error_local = NULL;
+		if (!fu_cfu_device_verify_descriptor(self, descriptor, &error_local)) {
+			if (descriptors_error->len > 0)
+				g_string_append(descriptors_error, ": ");
```
Comment (superm1) [inline plugins/cfu/fu-cfu-device.c:443]: Might be worth mentioning in the message which descriptor index has the errors to make debugging life easier in future

**Review by superm1 — APPROVED** [review]

File: libfwupdplugin/fu-hid-device.c (hunk 1 above)
Comment (hughsie) [reply libfwupdplugin/fu-hid-device.c:97]: Fixed (for 2.0.0)

File: plugins/cfu/fu-cfu-device.c (hunk 2 above)
Comment (hughsie) [reply plugins/cfu/fu-cfu-device.c:443]: Good idea, fixed.

**Review by superm1 — DISMISSED** [review]

**Pull request merged**